src/index.d.ts

```diff
@@ -75,3 +89,31 @@ export interface JsonRpcEngine {
     ) => void,
   ) => void;
 }
+
+export interface asMiddleware {
+  (engine: JsonRpcEngine): JsonRpcMiddleware;
+}
+
+export interface createAsyncMiddleware {
+  (asyncMiddleware: AsyncJsonrpcMiddleware): JsonRpcMiddleware;
+}
+
+type ScaffoldMiddlewareHandler = JsonRpcMiddleware | (
+  boolean | number | string | Record<string, unknown> | unknown[] | null | undefined
+)
```

Comment:
This one is a little iffy. Really, I want to type it like: _if_ it's a function, _then_ it has to extend `JsonRpcMiddleware`, otherwise it can be anything.

OTOH, even that is perhaps a bit too permissive, since the values should probably be serializable 🤔

Reply:
Maybe something like this? I think what you have is correct but just throwing out another possibility to look into.
```ts
type Serializable = boolean | number | string | Record<string, unknown> | unknown[] | null | undefined;
type ScaffoldMiddlewareHandler<T> = T extends Function ? JsonRpcMiddleware : Serializable;

export interface createScaffoldMiddleware<T> {
  (handlers: {[methodName: string]: ScaffoldMiddlewareHandler<T>}): JsonRpcMiddleware;
}
```